### DISCUSSION TITLE
Multiple audio latency issues caused by WASAPI minimum device buffer size

### DISCUSSION THREAD

**Discussion by Spok5508** [ppy/osu#29085] [Q&A]
### Type

Performance

### Bug description

I've been experiencing a lot of audio issues (offset, latency) in osu!lazer on windows, the biggest one being massive audio desync whenever you apply any rate changes (hit early on slowed down maps. hit late on sped up maps, both sometimes even up to +/- 50ms)

As the issue has been the biggest factor as to why I can't switch to lazer, I decided to look further into the subject and found out that [this](https://github.com/miniant-git/REAL) open-source project forces all audio streams to use a minimal buffer size based on your device's sample rate. When using this program, all of my audio issues are completely fixed.

I looked into the osu-framework source code and changed the device buffer size from 10ms to 3ms, with the exact same result. The audio feels what it should be. I also had no noticable increase of CPU usage when testing this.

Is this something that you guys would be willing to change? If so, I'd love to PR as I'm eager to contribute to the project.

### Screenshots or videos

[Link to the video on youtube where I demonstrate the issue](https://youtu.be/aI9f5TbyiYw?si=195Bs3gI6lZaSUWbl)

### Version

2024.725.0

### Logs

[compressed-logs.zip](https://github.com/user-attachments/files/16379538/compressed-logs.zip)

**Comment by bdach** [comment]
Minimal required reading for continuing this conversation:

- https://github.com/ppy/osu-framework/pull/6088, https://github.com/ppy/osu/pull/26154 (for what we planned for windows)
- https://github.com/ppy/osu-framework/pull/3823, https://github.com/ppy/osu-framework/pull/3832 (for demonstration of how messing with these arbitrarily will fix some devices but break others)
- https://github.com/ppy/osu-framework/pull/4008

**Comment by Spok5508** [comment]
> for what we planned for windows

Do these offset adjustments also address the occurrence of variable offset miss-timings in both -/+ relative to the rate adjustments? In [24015](https://github.com/ppy/osu/issues/24015) there's mentions of a hidden global offset that might help, but since slowed down audio is way earlier and sped up audio is late, I assumed that the underlying issue isn't some global offset based on user perception.

> for demonstration of how messing with these arbitrarily will fix some devices but break others

I'm aware that changing a random number until it feels good for me is not a good way to go about things, this issue was mostly created to hopefully give you guys some more insight into the audio issues from a player's perspective.

**Comment by bdach** [comment]
> since slowed down audio is way earlier and sped up audio is late, I assumed that the underlying issue isn't some global offset based on user perception

https://github.com/ppy/osu/discussions/27826 may be relevant here too then.

I'm not actually precisely sure why the WASAPI changes were shelved at this point. @peppy might know better.

**Comment by Spok5508** [comment]
I completely forgot to mention this, but rate changes actually line up if I disable exclusive audio mode under my default playback device. This is something that I discovered when researching WASAPI a bit more.

So instead of having random +50ms offset on DT, it just... works if I disable exclusive audio.

**Comment by peppy** [comment]
> I'm not actually precisely sure why the WASAPI changes were shelved at this point. @peppy might know better.

I was looking to improve shared mode, not use exclusive mode, but this didn't work for all people.

But none of this has anything to do with rate mods specifically.

**Reply by Spok5508** [reply]
Genuine question, if this is unrelated to rate mods, then how come all of my audio issues regarding rate mods are fixed if I disable exclusive mode?

**Reply by peppy** [reply]
I'm assuming you're using a third party application?

We don't support that, sorry.

**Reply by Spok5508** [reply]
Just unchecking this box in windows native settings 
![image](https://github.com/user-attachments/assets/22817ee4-eab3-4f24-9485-6be468056f69)

**Reply by Spok5508** [reply]
Just to reiterate, this checkbox is making the game go from having unplayable variable audio latency to none at all. I'm hoping it can give a clue to what is going on with all the audio issues.

**Reply by peppy** [reply]
yep thanks, will probably require more details from you soon
